### PR TITLE
iedorep command

### PR DESCRIPTION
**Pull request by bbdaniels**
`iedorep` ~initial~ version 0.2 for review. This version successfully scans a file in two runs and flags:

1. Changes in data
2. Data inconsistencies from Run 1 to Run 2 using `cf`
3. Use of RNG ("Seed Used")
4. Seed inconsistencies from Run 1 to Run 2 (when RNG state changes but end point differs)
5. Use of Sort RNG ("Sortseed Used")
6. Order inconsistencies after sort (combining `xpose` and `cf` whenever sort used)

Reporting is now as follows in the screenshot. Flags are displayed for all three types (whenever detected), as as errors (whenever inconsistencies are detected). ~Flags are not optional yet but can easily be done.~

Screen Shot(51)

Current syntax is: `iedorep /path/to/dofile.do`
(The `debug` option in the screenshot creates a copy of the underlying tempfile for debugging.)

🍀 Additional functionality possible for release:

- [ ] Clear user-written functions and set PERSONAL directory
- [ ] Create temp locations for results and hash all outputted files against 1st version

✳️ Core functionality:

- [x] Implement improvements from Luiza below
- [x] Implement pseudo-recursion
- [x] Write minimal help file
- [x] Set defaults and options
- [x] Implement error check for altered RNG
- [x] Implement `cf` check for altered data
- [x] Implement sort-check using data transpose
  - [x] Note that non-flag does not totally rule out error
  - [x] This checks for stability, but it's possible to get the same order by chance
  - [x] It also does not use string variables to check (limitation of `xpose`)

‼️ Critical needs:

- [x] Help file, error messages, optional functionality, etc.
- [x] Check path implementation in `syntax` for quote breakage
- [x] Implementation of recursive functionality so it can be run targeting a master do-file. I have some crude ideas but others will have better ideas.
- [x] Avoid using globals, or at least use better names for them (solved by using only one file)
  - [x] Make sure that `clear` will never cause error if run before file?
  - [x] Make sure that doubling the dofile will not cause errors with locals?

☺️ Nice-to-haves:

- [ ] ~Suppress `display` commands in target file~
- [ ] ~Suppress graph outputs in target file~
- [ ] ~Suppress exporting commands in target file?~

📝 Notes:

- There are almost certainly configurations of comments `/* */` and `{}` that will cause failures. This might be better addressed in `linter` than trying to catch all possible arrangements here.
- The command now compiles both runs of the dofile into a single tempfile (`cf` would not work otherwise, nor would using locals). It uses `clear` as well as advances the RNG state by one draw before each one, but this may be incorrect/insufficient. Especially counter locals may break. Thinking of ways to catch this.
- ~I cannot figure out how to detect changes in the RNG state from one run to the next. Currently it flags whenever the RNG state is changed so those lines can be checked.~
- Loops aren't evaluated internally; it currently treats the highest level loop as a single line. This functionality needs to be checked.
- ~Data flagging is probably too verbose; it flags on every line after the first difference in many cases (since the differences persist). This could be changed easily to simply break on the first failure.~

### CODE REVIEW

File: src/ado_files/iedorep.ado [hunk 1]
```diff
@@ -0,0 +1,92 @@
+*! version 0.1 9NOV2021 DIME Analytics dimeanalytics@worldbank.org
+
+cap  program drop  iedorep
+  program define   iedorep, rclass
+  
+  syntax anything
+  
+// Prep
+  file close _all
```
Comment (kbjarkefur) [inline src/ado_files/iedorep.ado:9]: Is this only closing the files open in this scope? Or is it closing also some other file that still should be open outside this command? Either only close the only file needed to be closed or carefully test the scope of `_all` here.

**Comment by luizaandrade** [comment]
- [ ] Only skip lines with `{` if they also include the expressions "foreach", "while" or "forvalues" and its shortcuts.
- [ ] Allow for quotation marks in file paths

File: src/help_files/iedorep.sthlp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  and sort seed state, regardless of whether they are flagged as errors.
+
+{marker optslong}
+{title:Options}
+
+{phang}The {cmdab:recursive} option requests that {cmdab:iedorep} detect when errors
+  are flagged following a line that uses {cmdab:do} or {cmdab:run}
+  to execute "sub" do-files. If such errors are detected (ie, if the sub-do-file
+  leaves the data, RNG state, or sort order different in the second run),
+  {cmdab:iedorep} will execute again targeting that filepath.
```
Comment (luisesanmartin) [inline src/help_files/iedorep.sthlp:61]: I didn't understand that without the `recursive` option `iedorep` still executes sub do-files but just treats the execution line as any other, I thought that without it the command would just skip those lines. I suggest adding something like this at the end of this paragraph:
> If **recursive** is not used, **iedorep** will still run sub do-files as part of the lines where **do** or **run** are used.

**Review by luisesanmartin — COMMENTED** [review]
Thanks a lot, @bbdaniels. All the other issues I noticed in my last review are solved, they are closed now.

This is what I'm adding in this review:
- Suggested replacing "replicability" with "reproducibility" in the help file
- Added a couple more suggestions in the help file so it's clearer for new users
- There's a possible bug described in #304 that should be reviewed before releasing v1.0. After that, everything should be in place for the first release

File: src/help_files/iedorep.sthlp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  
+{phang}{cmdab:iedorep} will run the do-file twice and flag three types of possible
+  errors between the first and the second run of the do-file.
+  (1) Changes to data which leave the data at a different state 
+  in the same line of the do-file. (Changes in the datasignature)
+  (2) Changes to the random number generator state 
+  which result in a different state at the same line in the do-file. (Changes in the RNG state)
+  (3) Changes to the sort order of the data which leave the data
+  in a different sort order at the same line of the do-file. (Changes in the sortseed)
+  {break}
```
[the remark this answers is not in the document]
Comment (bbdaniels) [reply src/help_files/iedorep.sthlp]: Accepted and implemented.

File: src/help_files/iedorep.sthlp (hunk 2 above)
Comment (bbdaniels) [reply src/help_files/iedorep.sthlp:61]: Done in d3b213145552dae288c540084b02b8d210b01bb0

**Comment by bbdaniels** [comment]
@luisesanmartin Great comments, all resolved except #304. Have a look at the proposed solution there and let me know if that's sufficient to go ahead!

File: src/help_files/iedorep.sthlp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+{title:Title}
+
+{phang}{cmdab:iedorep} {hline 2} Returns information on the version of iedorep installed
+
+{phang}For a more descriptive discussion on the intended usage and work flow of this
+command please see the {browse "https://dimewiki.worldbank.org/wiki/iedorep":DIME Wiki}.
+
+{title:Syntax}
+
+{phang2} {cmdab:iedorep} "{it:/path/do/do-file.do}" , [ {cmdab:r:ecursive} {cmdab:alldata allseed allsort} ]
```
Comment (kbjarkefur) [inline src/help_files/iedorep.sthlp:17]:
```suggestion
{phang2} {cmdab:iedorep} "{it:/path/to/do-file.do}" , [ {cmdab:r:ecursive} {cmdab:alldata allseed allsort} ]
```

File: src/ado_files/iedorep.ado [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+        iedorep `file' , `debug' `qui' `recursive' `alldata' `allsort' `allseed'
+        if `r(errors)' ==  0 {
+          di as err "!!------ WARNING ------!!"
+          di as err "An error was expected in this sub do-file but none was found."
+          di as err "  This can occur when sub do-files are not independent."
+          di as err "  Check that this sub do-file does not depend on the calling file."
+          di as err "  This includes both the data state and global macros;"
+          di as err "    the data is reset on each run, but macros are not."
+          di as err " "
+          error 459
```
Comment (kbjarkefur) [inline src/ado_files/iedorep.ado:433]: Suggest printing this using `smcl` such that spacing and line breaks works as expected no matter the screen width. This seems like a quick fix that will look bad.

File: src/help_files/iedorep.sthlp [hunk 6]
```diff
@@ -0,0 +1,136 @@
+{smcl}
+{* 11 Jan 2022}{...}
+
+{hline}
+help for {hi:iedorep}
+{hline}
+
+{title:Title}
+
+{phang}{cmdab:iedorep} {hline 2} Returns information on the version of iedorep installed
```
Comment (bbdaniels) [inline src/help_files/iedorep.sthlp:10]:
```suggestion
{phang}{cmdab:iedorep} {hline 2} detects reproducibility errors in do-files automatically by running them multiple times and comparing the Stata state on a line-by-line basis between executions. 
```

File: src/ado_files/iedorep.ado [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+/*****************************************************************************
+Pseudo-recursion
+*****************************************************************************/
+
+  if "`recursive'" != "" {
+    qui keep if !(Err_1 == "" & Err_2 == "" & Err_3 == "") & (Path != "")
+    if `c(N)' == 0 {
+      di as err " "
+      di as err "No errors detected in sub do-files; recursion halted."
```
Comment (bbdaniels) [inline src/ado_files/iedorep.ado:413]:
```suggestion
      di as err "No errors detected in sub do-files; recursion completed."
```

[the remark this answers is not in the document]
Comment (bbdaniels) [reply src/ado_files/iedorep.ado]: Changed to "completed".

File: src/ado_files/iedorep.ado (hunk 5 above)
Comment (bbdaniels) [reply src/ado_files/iedorep.ado:433]: It is, but as you correctly point out in https://github.com/worldbank/ietoolkit/issues/304#issuecomment-1342681506, it's temporary anyway. Can we go ahead and publish as is?

**Review by luisesanmartin — APPROVED** [review]
Thanks a lot, @bbdaniels . This is ready to be published as v1.
Note to ourselves: further work for later versions are described in #304, #306, and #307

**Pull request merged**